### ISSUE TITLE
possible race condition with set_logger (reproducible)

### ISSUE THREAD

**Issue by jberkenbilt** [rust-lang/log#650]
Tl;Dr: attached is an example program that, when run repeatedly, occasionally fails in a way that suggests that `set_logger` never called the initialization function. Synchronizing the calls to `log::set_logger` makes the problem non-reproducible. Details below.

I have implemented a test logger that captures log messages to a Vec. It uses `thread_local!` to ensure that each thread has its own buffer so it's safe to use in tests. Once in a while, a test fails because no log data has been captured. Upon investigating, I have observed that protecting the call to `log::set_logger` with a `LazyLock` appears to remove this problem. Looking through the code, I see that, at least on my platform, `set_logger` ultimately calls a function that uses an atomic usize to protect code that uses `unsafe` to initialize a mutable static. While I realize it has impact on MSRV, with current rust (I think 1.80), putting the global logger in a `LazyLock` would seem less prone to race condition. I can't prove that there is definitely a race condition in the existing implementation, but the pattern of using an atomic to protect _something else_ feels a lot more race-condition prone than using a `LazyLock`. I was thinking of doing a pull request, but I wasn't sure about what conditional compilation would be desired for ensuring that the `LazyLock` solution was only used on platforms and rust versions that could support it. (A `std::sync::Once` would probably also work, and that's been around longer.)

To reproduce, extract the attached tar file (which contains just Cargo.toml and src/lib.rs). Then
- run `cargo test`
- get name of test executable from output, e.g. target/debug/deps/logging-670093953cafad50
- run test repeatedly until it fails: `while ./target/debug/deps/logging-670093953cafad50; do true; done`

[set-logger.tar.gz](https://github.com/user-attachments/files/18165710/set-logger.tar.gz)

On my macbook pro running the latest version of the OS and the latest stable rust, it usually takes about 10 seconds before the bug triggers. When I run this using an arm64 Ubuntu 24.04 in a `lima` VM on the same hardware, it usually reproduces a little faster. I have noticed that wrapping the call to `set_logger` itself in a `LazyLock` seems to avoid the problem. A workaround is commented out in the example with `//>`. Unfortunately, I was not able to run the test for an infinite amount of time to be 100% sure, but I ran it for several minutes when it usually fails within just a few seconds. I was also able to reproduce the issue on my native Linux box, which has a Ryzen 9 and is running Ubuntu 24.04, also with the latest rust. The attached code is the minimal version I could create that still reliably fails with repeated runs.

```rust
// To reproduce bug:
// - run `cargo test`
// - get name of test executable from output, e.g. target/debug/deps/logging-670093953cafad50
// - run test repeatedly: `while ./target/debug/deps/logging-670093953cafad50; do true; done`

use log::{LevelFilter, Log, Metadata, Record};
use std::cell::RefCell;

static TEST_LOGGER: &TestLogger = &TestLogger;
pub struct TestLogger;

impl TestLogger {
    fn with_local_messages_mut<F, Ret>(f: F) -> Ret
    where
        F: FnMut(&mut Option<Vec<String>>) -> Ret,
    {
        thread_local! {
            static MESSAGES: RefCell<Option<Vec<String>>> = const { RefCell::new(None) };
        }
        MESSAGES.with_borrow_mut(f)
    }

    pub fn init(level: LevelFilter) {
        if log::set_logger(TEST_LOGGER).is_ok() {
            log::set_max_level(level);
        }
    }

    pub fn messages() -> Vec<String> {
        Self::with_local_messages_mut(|m| m.take().unwrap_or_default())
    }
}

impl Log for TestLogger {
    fn enabled(&self, _metadata: &Metadata) -> bool {
        true
    }
    fn log(&self, record: &Record) {
        if self.enabled(record.metadata()) {
            Self::with_local_messages_mut(|messages| {
                if messages.is_none() {
                    *messages = Some(Vec::new());
                }
                messages.as_mut().unwrap().push(record.args().to_string());
            });
        }
    }
    fn flush(&self) {}
}

#[cfg(test)]
mod tests {
    use super::*;
    //> use std::sync::LazyLock;

    fn init_logger() {
        // Uncommenting the LazyLock logic works around the issue.

        //> static LOG: LazyLock<()> = LazyLock::new(|| {
        TestLogger::init(LevelFilter::Info);
        //> });
        //> *LOG;
    }

    #[test]
    fn test_logger1() {
        init_logger();
        log::info!("asdf");
        log::warn!("qwer");
        assert_eq!(
            TestLogger::messages(),
            vec!["asdf".to_string(), "qwer".to_string()]
        );
        assert_eq!(TestLogger::messages().len(), 0);
    }

    #[test]
    fn test_logger2() {
        init_logger();
        log::error!("moo");
        assert_eq!(TestLogger::messages(), vec!["moo".to_string()]);
        assert_eq!(TestLogger::messages().len(), 0);
    }
}
```

**Comment by sfackler** [comment]
The race condition is in your init function:

```rust
    pub fn init(level: LevelFilter) {
        if log::set_logger(TEST_LOGGER).is_ok() {
            log::set_max_level(level);
        }
    }
```

When called concurrently, the thread that loses the race to call set_logger first can return before set_max_level has been called.

**Comment by jberkenbilt** [comment]
Yes, I see now. Thanks.
